File: src/main/kotlin/io/titandata/titan/commands/Push.kt
/*
 * Copyright (c) 2019 by Delphix. All rights reserved.
 */

package io.titandata.titan.commands

import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.core.requireObject
import com.github.ajalt.clikt.parameters.arguments.argument
import com.github.ajalt.clikt.parameters.options.flag
import com.github.ajalt.clikt.parameters.options.multiple
import com.github.ajalt.clikt.parameters.options.option
import io.titandata.titan.Dependencies
import org.kodein.di.Kodein
import org.kodein.di.generic.bind
import org.kodein.di.generic.inSet
import org.kodein.di.generic.provider

class Push : CliktCommand(help = "Push data state to remote") {
    private val dependencies: Dependencies by requireObject()
    private val commit: String? by option("-c", "--commit", help = "Commit GUID to push, defaults to latest")
    private val remote: String? by option("-r", "--remote", help = "Name of the remote provider, defaults to origin")
    private val tags by option("-t", "--tag", help = "Filter commits to select commit to push").multiple()
    private val metadataOnly: Boolean by option("-u", "--update-only", help = "Update tags only, do not push data").flag()
    private val repository: String by argument()

    override fun run() {
        val (provider, repoName) = dependencies.providers.byRepository(repository)
        provider.push(repoName, commit, remote, tags, metadataOnly)
    }
}

val pushModule = Kodein.Module("push") {
    bind<CliktCommand>().inSet() with provider {
        Push()
    }
}

File: internal/app/commands/push.go
/*
Copyright © 2019 The Titan Project Contributors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
package commands

import (
	"github.com/spf13/cobra"
)

// pushCmd represents the push command
var pushCmd = &cobra.Command{
	Use:   "push [REPOSITORY]",
	Short: "Push data state to remote",
	Args: cobra.ExactArgs(1),
	Run: func(cmd *cobra.Command, args []string) {
		repo := args[0]
		provider.Push(repo, guid, remote, tags, updateOnly)
	},
}

func init() {
	rootCmd.AddCommand(pushCmd)
	pushCmd.Flags().StringVarP(&guid, "commit","c", "","commit GUID to push, defaults to latest")
	pushCmd.Flags().StringVarP(&remote, "remote","r", "","name of the remote provider, defaults to origin")
	pushCmd.Flags().BoolVarP(&updateOnly, "update-only", "u", false, "update tags only, do not push data")
	pushCmd.Flags().StringSliceVarP(&tags, "tags", "t", nil, "filter commits to select commit to push")
	pushCmd.Flags().SortFlags = false //TODO review flag sorting
}
